### PR TITLE
HRQB 21 - Employees table

### PR DESCRIPTION
**Pull request by ghukill**
### Purpose and background context

This PR introduces the first tasks that will load data in Quickbase, specifically the `Employees` table.  As a foundational table, this made sense to start with, and it's relatively simple with no dependencies.

As the first table to be loaded, it needed to be situated in a "pipeline" in the HRQBClient.  While we may get into incremental updates over time, the first pipeline established is a full refresh/upsert called `FullUpdate`.

It's gauranteed that some columns will shift, new `Employee` fields will be added, but this establishes a baseline to work with while other tables will follow shortly.

The goal of this PR is to knowledge share the basic structure of ETL tasks for loading data into a single table, and futures ones will follow _very_ similar patterns.

Testing is still minimal for the tasks themselves.  The base tasks, which these extend, are well tested.  ~~A meaningful test at this level would be against the data itself, where mocking it begins to present a questionable return-on-investment for what the test provides.  I'm proposing to skip tests for this PR, to focus on the mechanics of a pipeline, but aim to revisit them in future PRs.~~  Some preliminary tests have been created, with the assumption that future tests may be added as other QB table tasks take shape and expose edge cases to test for.

### How can a reviewer manually see the effects of these changes?

Still figuring out sensible access restrictions, so at this time direct testing is not possible.  But can demonstrate some commands and output.

Get status of pipeline:
```shell
pipenv run hrqb --verbose pipeline -p FullUpdate status
```

Output:
```
2024-05-28 14:09:08,827 INFO hrqb.cli.main() line 31: Logger 'root' configured with level=DEBUG
2024-05-28 14:09:08,827 INFO hrqb.cli.main() line 32: No Sentry DSN found, exceptions will not be sent to Sentry
2024-05-28 14:09:08,827 INFO hrqb.cli.main() line 34: Running process
2024-05-28 14:09:08,832 DEBUG hrqb.cli.pipeline() line 123: Successfully loaded pipeline: 'hrqb.tasks.pipelines.FullUpdate'
2024-05-28 14:09:08,834 INFO hrqb.cli.status() line 143: 
├── INCOMPLETE: FullUpdate()
   ├── INCOMPLETE: LoadEmployees(pipeline=FullUpdate, table_name=Employees, stage=Load)
      ├── INCOMPLETE: TransformEmployees(pipeline=FullUpdate, table_name=, stage=Transform)
         ├── INCOMPLETE: ExtractDWEmployees(table_name=, pipeline=FullUpdate, stage=Extract)
```
- note the root `FullUpdate` task
- not the ETL tasks that support loading the `Employees` table

Running the pipeline:
```shell
pipenv run hrqb --verbose pipeline -p FullUpdate run
```

Output:
```
...
...
2024-05-28 14:10:21,158 INFO hrqb.base.task.parse_and_log_upsert_results() line 255: Record processed: 751, created: 0, modified: 0, unchanged: 751
...
...
===== Luigi Execution Summary =====

Scheduled 4 tasks of which:
* 4 ran successfully:
    - 1 ExtractDWEmployees(table_name=, pipeline=FullUpdate, stage=Extract)
    - 1 FullUpdate(parent_pipeline_name=)
    - 1 LoadEmployees(pipeline=FullUpdate, table_name=Employees, stage=Load)
    - 1 TransformEmployees(pipeline=FullUpdate, table_name=, stage=Transform)

This progress looks :) because there were no failed tasks or missing dependencies

===== Luigi Execution Summary =====

2024-05-28 14:10:21,400 INFO hrqb.cli.run() line 192: Pipeline run result: SUCCESS
2024-05-28 14:10:21,401 INFO hrqb.cli.run() line 193: 
├── COMPLETE: FullUpdate()
   ├── COMPLETE: LoadEmployees(pipeline=FullUpdate, table_name=Employees, stage=Load)
      ├── COMPLETE: TransformEmployees(pipeline=FullUpdate, table_name=, stage=Transform)
         ├── COMPLETE: ExtractDWEmployees(table_name=, pipeline=FullUpdate, stage=Extract)
```
- because data was already loaded, no records were modified (`unchanged: 751` in logs)
  - demonstrating that it's idempotent upserts, for this table at least
- when complete, the entire `FullUpdate` pipeline is considered complete; only a single task so far, but others will follow

### Includes new or updated dependencies?
YES

### Changes expectations for external applications?
NO

### What are the relevant tickets?
- https://mitlibraries.atlassian.net/browse/HRQB-21

### Developer
- [X] All new ENV is documented in README
- [X] All new ENV has been added to staging and production environments
- [X] All related Jira tickets are linked in commit message(s)
- [X] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed **or** provided examples verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

File: hrqb/tasks/pipelines.py [hunk 1]
```diff
@@ -1,3 +1,17 @@
...
+import luigi  # type: ignore[import-untyped]
+
+from hrqb.base.task import HRQBPipelineTask
+
+
+class FullUpdate(HRQBPipelineTask):
+    """Pipeline to perform a full update of all Quickbase tables."""
+
+    def requires(self) -> Iterator[luigi.Task]:  # pragma: no cover
+        from hrqb.tasks.employees import LoadEmployees
```
Comment (ghukill) [inline hrqb/tasks/pipelines.py:14]: Imports are performed inside the `requires()` method to avoid them imported at the top level.  This keeps imports from _this_ file limited to `HRQBPipelineTask` tasks themselves.

File: hrqb/tasks/pipelines.py [hunk 2]
```diff
@@ -1,3 +1,17 @@
...
+
+        yield LoadEmployees(pipeline=self.pipeline_name)
```
Comment (ghukill) [inline hrqb/tasks/pipelines.py:16]: As more table ETL tasks are added, the final load task will get added here.  I would estimate this will be 10-12 tasks long by the end.

File: hrqb/base/task.py [hunk 3]
```diff
@@ -210,14 +230,30 @@ def run(self) -> None:
...
+            return
+        processed = metadata.get("totalNumberOfRecordsProcessed", 0)
+        created = len(metadata.get("createdRecordIds", []))
+        updated = len(metadata.get("updatedRecordIds", []))
+        unchanged = len(metadata.get("unchangedRecordIds", []))
+        message = (
+            f"Record processed: {processed}, created: {created}, "
+            f"modified: {updated}, unchanged: {unchanged}"
+        )
+        logger.info(message)
```
Comment (ehanson8) [inline hrqb/base/task.py:255]: Elegant response parsing!

File: hrqb/utils/__init__.py [hunk 4]
```diff
@@ -3,15 +3,52 @@
...
+from dateutil.parser import parse as date_parser  # type: ignore[import-untyped]
 
 
 def today_date() -> datetime.date:
     return datetime.datetime.now(tz=datetime.UTC).date()
 
 
+def normalize_date(date: str | datetime.datetime) -> str | None:
+    if date is None or date == "" or pd.isna(date):
+        return None
```
Comment (ehanson8) [inline hrqb/utils/__init__.py:21]: Could this just be `pass` since the default is to return `None`?

File: hrqb/utils/__init__.py [hunk 5]
```diff
@@ -3,15 +3,52 @@
...
+def convert_oracle_bools_to_qb_bools(
+    dataframe: pd.DataFrame,
+    columns: list[str],
+) -> pd.DataFrame:
+    """Convert Oracle bools [Y,N] to Quickbase bools [Yes,No] for a subset of columns."""
+    dataframe[columns] = dataframe[columns].replace({"Y": "Yes", "N": "No"})
+    return dataframe
+
+
+def state_abbreviation_to_name(abbreviation: str | None) -> str | None:
```
Comment (ehanson8) [inline hrqb/utils/__init__.py]: Optional and minor: you might update the name or adding a docstring to clarify this is about US states and not CS concept "state"

File: hrqb/tasks/employees.py
Comment (ehanson8) [inline hrqb/tasks/employees.py:1]: Impressed at how clean this is, your infrastructure development really pays off when the tasks are so easy to read. Great work!

**Review by ehanson8 — COMMENTED** [review]
Looks great, a few notes

File: hrqb/base/task.py (hunk 3 above)
Comment (ghukill) [reply hrqb/base/task.py:255]: Ah thanks!  At the time I was tired and just wanted it to be there, so it's right down the middle boring.  And, now, a good reminder that that kind of code is usually a good bet.

File: hrqb/utils/__init__.py (hunk 4 above)
Comment (ghukill) [reply hrqb/utils/__init__.py:21]: If you don't mind, I kind of prefer the explicit `return None`. I think just `return` (with an implied `None`) is okay, but less explicit for code scanning.  

As for `pass`, I consider that more for placeholders, or when you have a conditional block that you're force to have for some reason, but don't want anything to happen there.  

In this case, just feels like a default, fall-through return value.

File: hrqb/utils/__init__.py (hunk 5 above)
Comment (ghukill) [reply hrqb/utils/__init__.py]: Good call!  Will do.

File: hrqb/tasks/employees.py
Comment (ghukill) [reply hrqb/tasks/employees.py:1]: Thanks!  Appreciate it.  Time will tell, but hopefully there will be a benefit to the base tasks having the complexity, and these just focusing on fields and values.

**Review by ehanson8 — APPROVED** [review]

File: hrqb/cli.py [hunk 6]
```diff
@@ -72,7 +72,7 @@ def test_connections(ctx: click.Context) -> None:
     )
 
 
-@click.group()
+@main.group()
```
Comment (jonavellecuerdo) [inline hrqb/cli.py:75]: [Non-blocking] Does your IDE (PyCharm) still throw a linting error for this syntax? 🤔

File: hrqb/utils/__init__.py [hunk 7]
```diff
@@ -3,15 +3,52 @@
...
+import us  # type: ignore[import-untyped]
+from dateutil.parser import parse as date_parser  # type: ignore[import-untyped]
 
 
 def today_date() -> datetime.date:
     return datetime.datetime.now(tz=datetime.UTC).date()
 
 
+def normalize_date(date: str | datetime.datetime) -> str | None:
+    if date is None or date == "" or pd.isna(date):
```
Comment (jonavellecuerdo) [inline hrqb/utils/__init__.py]: Minor but [`pd.isna` also returns True for `None`](https://pandas.pydata.org/docs/reference/api/pandas.isna.html#pandas.isna).

File: hrqb/utils/__init__.py [hunk 8]
```diff
@@ -3,15 +3,52 @@
...
+        return None
+    if isinstance(date, str):
+        date = date_parser(date)
+    if isinstance(date, datetime.datetime):
+        return date.strftime("%Y-%m-%d")
+    return None
```
Comment (jonavellecuerdo) [inline hrqb/utils/__init__.py]: Hmm, when viewing this code locally, `Pylance` indicates "Code is unreachable" for the final `return None`. Should there be some error handling for dates that cannot be parsed?

Perhaps rewrite to:

```
import logging

def normalize_date(date: str | datetime.datetime) -> str | None:
    if date is None or date == "" or pd.isna(date):
        return None
    if isinstance(date, str):
        try:
            date = date_parser(date)
        except ParserError:
            logging.warning("Unable to parse date from '{date}'")
    if isinstance(date, datetime.datetime):
        return date.strftime("%Y-%m-%d")
    return None
```

File: hrqb/tasks/employees.py
Comment (jonavellecuerdo) [reply hrqb/tasks/employees.py:1]: I agree 💯 !

**Review by jonavellecuerdo — COMMENTED** [review]
This is looking really great! Just have a few comments for you, @ghukill.

File: hrqb/cli.py (hunk 6 above)
Comment (ghukill) [reply hrqb/cli.py:75]: Not here, but I get an IDE linting error for this usage:

```python
@pipeline.command()
...
```

I'm okay with it.  Things work as expected, and mypy and ruff are fine; seems to just be my IDE.

File: hrqb/utils/__init__.py (hunk 7 above)
Comment (ghukill) [reply hrqb/utils/__init__.py]: Nice catch!  Will remove that redundant `is None` clause.

File: hrqb/utils/__init__.py (hunk 8 above)
Comment (ghukill) [reply hrqb/utils/__init__.py]: This utility function has been a headache, and @ehanson8 commented on it too.  I think this is a good change.  You're (and pylance) are right, that if the string is not parsable, it would raise an exception so we'd never hit that `return None` (because if it _did_ parse it, then we'd return on the next block).  Updating!

**Comment by ghukill** [comment]
New commit to address comments @jonavellecuerdo

**Review by jonavellecuerdo — APPROVED** [review]
